File: src/main/ioke/lang/FileSystem.java
/*
 * See LICENSE file in distribution for copyright and licensing information.
 */
package ioke.lang;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.BufferedReader;
import java.io.Writer;

import java.nio.channels.FileChannel;

import java.util.ArrayList;
import java.util.List;
import java.util.HashMap;
import java.util.Map;

import ioke.lang.exceptions.ControlFlow;

import ioke.lang.util.Dir;

/**
 *
 * @author <a href="mailto:ola.bini@gmail.com">Ola Bini</a>
 */
public class FileSystem {
    public static List<String> glob(Runtime runtime, String text) {
        return Dir.push_glob(runtime.getCurrentWorkingDirectory(), text, 0);
    }

    public static class IokeFile extends IokeIO {
        private File file;

        public IokeFile(File file) {
            super(null, null);
            this.file = file;

            try {
                if(file != null) {
                    this.writer = new FileWriter(file);
                }
            } catch(IOException e) {
            }
        }

        @Override
        public void init(IokeObject obj) throws ControlFlow {
            final Runtime runtime = obj.runtime;

            obj.setKind("FileSystem File");

            obj.registerMethod(runtime.newJavaMethod("Closes any open stream to this file", new TypeCheckingJavaMethod.WithNoArguments("close", obj) {
                @Override
                public Object activate(IokeObject method, Object on, List<Object> args, Map<String, Object> keywords, IokeObject context, IokeObject message) throws ControlFlow {
                    try {
                        Writer writer = IokeFile.getWriter(on);
                        if(writer != null) {
                             writer.close();
                        }
                    } catch(IOException e) {
                    }
                    return context.runtime.nil;
                }
            }));
        }
    }

    public static void init(IokeObject obj) throws ControlFlow {
        Runtime runtime = obj.runtime;
        obj.setKind("FileSystem");

        IokeObject file = new IokeObject(runtime, "represents a file in the file system", new IokeFile(null));
        file.mimicsWithoutCheck(runtime.io);
        file.init();
        obj.registerCell("File", file);

        obj.registerMethod(runtime.newJavaMethod("Tries to interpret the given arguments as strings describing file globs, and returns an array containing the result of applying these globs.", new JavaMethod("[]") {
                private final DefaultArgumentsDefinition ARGUMENTS = DefaultArgumentsDefinition
                    .builder()
                    .withRest("globTexts")
                    .getArguments();

                @Override
                public DefaultArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject method, IokeObject context, IokeObject message, Object on) throws ControlFlow {
                    List<Object> args = new ArrayList<Object>();
                    getArguments().getEvaluatedArguments(context, message, on, args, new HashMap<String, Object>());
                    List<String> dirs = FileSystem.glob(context.runtime, Text.getText(args.get(0)));
                    List<Object> result = new ArrayList<Object>();
                    for(String s : dirs) {
                        result.add(context.runtime.newText(s));
                    }
                    return context.runtime.newList(result);
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("Takes one string argument and returns true if it's the relative or absolute name of a directory, and false otherwise.", new JavaMethod("directory?") {
                private final DefaultArgumentsDefinition ARGUMENTS = DefaultArgumentsDefinition
                    .builder()
                    .withRequiredPositional("directoryName")
                    .getArguments();

                @Override
                public DefaultArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject method, IokeObject context, IokeObject message, Object on) throws ControlFlow {
                    List<Object> args = new ArrayList<Object>();
                    getArguments().getEvaluatedArguments(context, message, on, args, new HashMap<String, Object>());

                    String name = Text.getText(args.get(0));
                    File f = null;
                    if(IokeSystem.isAbsoluteFileName(name)) {
                        f = new File(name);
                    } else {
                        f = new File(context.runtime.getCurrentWorkingDirectory(), name);
                    }
                    
                    return f.isDirectory() ? context.runtime._true : context.runtime._false;
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("Takes one string argument that should be a file name, and returns a text of the contents of this file.", new JavaMethod("readFully") {
                private final DefaultArgumentsDefinition ARGUMENTS = DefaultArgumentsDefinition
                    .builder()
                    .withRequiredPositional("fileName")
                    .getArguments();

                @Override
                public DefaultArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject method, IokeObject context, IokeObject message, Object on) throws ControlFlow {
                    List<Object> args = new ArrayList<Object>();
                    getArguments().getEvaluatedArguments(context, message, on, args, new HashMap<String, Object>());

                    String name = Text.getText(args.get(0));
                    File f = null;
                    if(IokeSystem.isAbsoluteFileName(name)) {
                        f = new File(name);
                    } else {
                        f = new File(context.runtime.getCurrentWorkingDirectory(), name);
                    }
                    StringBuilder sb = new StringBuilder();

                    try {
                        BufferedReader reader = new BufferedReader(new FileReader(f));
                        char[] buf = new char[1024];
                        int read = -1;
                        while((read = reader.read(buf, 0, 1024)) != -1) {
                            sb.append(buf, 0, read);
                        }
                        reader.close();
                    } catch(IOException e) {
                    }
                    
                    return context.runtime.newText(sb.toString());
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("Takes one string argument and returns true if it's the relative or absolute name of a file, and false otherwise.", new JavaMethod("file?") {
                private final DefaultArgumentsDefinition ARGUMENTS = DefaultArgumentsDefinition
                    .builder()
                    .withRequiredPositional("fileName")
                    .getArguments();

                @Override
                public DefaultArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject method, IokeObject context, IokeObject message, Object on) throws ControlFlow {
                    List<Object> args = new ArrayList<Object>();
                    getArguments().getEvaluatedArguments(context, message, on, args, new HashMap<String, Object>());

                    String name = Text.getText(args.get(0));
                    File f = null;
                    if(IokeSystem.isAbsoluteFileName(name)) {
                        f = new File(name);
                    } else {
                        f = new File(context.runtime.getCurrentWorkingDirectory(), name);
                    }
                    
                    return f.isFile() ? context.runtime._true : context.runtime._false;
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("Takes one string argument and returns true if it's the relative or absolute name of something that exists.", new JavaMethod("exists?") {
                private final DefaultArgumentsDefinition ARGUMENTS = DefaultArgumentsDefinition
                    .builder()
                    .withRequiredPositional("entryName")
                    .getArguments();

                @Override
                public DefaultArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject method, IokeObject context, IokeObject message, Object on) throws ControlFlow {
                    List<Object> args = new ArrayList<Object>();
                    getArguments().getEvaluatedArguments(context, message, on, args, new HashMap<String, Object>());

                    String name = Text.getText(args.get(0));
                    File f = null;
                    if(IokeSystem.isAbsoluteFileName(name)) {
                        f = new File(name);
                    } else {
                        f = new File(context.runtime.getCurrentWorkingDirectory(), name);
                    }
                    
                    return f.exists() ? context.runtime._true : context.runtime._false;
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("Takes one string argument that should be the path of a file or directory, and returns the parent of it - or nil if there is no parent.", new JavaMethod("parentOf") {
                private final DefaultArgumentsDefinition ARGUMENTS = DefaultArgumentsDefinition
                    .builder()
                    .withRequiredPositional("entryName")
                    .getArguments();

                @Override
                public DefaultArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject method, IokeObject context, IokeObject message, Object on) throws ControlFlow {
                    List<Object> args = new ArrayList<Object>();
                    getArguments().getEvaluatedArguments(context, message, on, args, new HashMap<String, Object>());

                    String name = Text.getText(args.get(0));
                    File f = null;
                    if(IokeSystem.isAbsoluteFileName(name)) {
                        f = new File(name);
                    } else {
                        f = new File(context.runtime.getCurrentWorkingDirectory(), name);
                    }
                    
                    String parent = f.getParent();
                    if(parent == null) {
                        return context.runtime.nil;
                    }

                    String cwd = context.runtime.getCurrentWorkingDirectory();

                    if(!IokeSystem.isAbsoluteFileName(name) && parent.equals(cwd)) {
                        return context.runtime.nil;
                    }

                    if(parent.startsWith(cwd)) {
                        parent = parent.substring(cwd.length()+1);
                    }

                    return context.runtime.newText(parent);
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("Takes a file name and a lexical block - opens the file, ensures that it exists and then yields the file to the block. Finally it closes the file after the block has finished executing, and then returns the result of the block.", new JavaMethod("withOpenFile") {
                private final DefaultArgumentsDefinition ARGUMENTS = DefaultArgumentsDefinition
                    .builder()
                    .withRequiredPositional("fileName")
                    .withRequiredPositional("code")
                    .getArguments();

                @Override
                public DefaultArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject method, IokeObject context, IokeObject message, Object on) throws ControlFlow {
                    List<Object> args = new ArrayList<Object>();
                    getArguments().getEvaluatedArguments(context, message, on, args, new HashMap<String, Object>());

                    String name = Text.getText(args.get(0));
                    File f = null;
                    if(IokeSystem.isAbsoluteFileName(name)) {
                        f = new File(name);
                    } else {
                        f = new File(context.runtime.getCurrentWorkingDirectory(), name);
                    }

                    try {
                        if(!f.exists()) {
                            f.createNewFile();
                        }
                    } catch(IOException e) {
                    }

                    IokeObject ff = context.runtime.newFile(context, f);
                    Object result = context.runtime.nil;

                    try {
                        result = context.runtime.callMessage.sendTo(context, args.get(1), ff);
                    } finally {
                        context.runtime.closeMessage.sendTo(context, ff);
                    }

                    return result;
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("Copies a file. Takes two text arguments, where the first is the name of the file to copy and the second is the name of the destination. If the destination is a directory, the file will be copied with the same name, and if it's a filename, the file will get a new name", new JavaMethod("copyFile") {
                private final DefaultArgumentsDefinition ARGUMENTS = DefaultArgumentsDefinition
                    .builder()
                    .withRequiredPositional("fileName")
                    .withRequiredPositional("destination")
                    .getArguments();

                @Override
                public DefaultArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject method, IokeObject context, IokeObject message, Object on) throws ControlFlow {
                    List<Object> args = new ArrayList<Object>();
                    getArguments().getEvaluatedArguments(context, message, on, args, new HashMap<String, Object>());

                    String name = Text.getText(args.get(0));
                    File f = null;
                    if(IokeSystem.isAbsoluteFileName(name)) {
                        f = new File(name);
                    } else {
                        f = new File(context.runtime.getCurrentWorkingDirectory(), name);
                    }

                    String name2 = Text.getText(args.get(1));
                    File f2 = null;
                    if(IokeSystem.isAbsoluteFileName(name2)) {
                        f2 = new File(name2);
                    } else {
                        f2 = new File(context.runtime.getCurrentWorkingDirectory(), name2);
                    }

                    if(f2.isDirectory()) {
                        f2 = new File(f2, f.getName());
                    }


                    try {
                        if(!f2.exists()) {
                            f2.createNewFile();
                        }

                        FileChannel srcChannel = new FileInputStream(f).getChannel();
                        FileChannel dstChannel = new FileOutputStream(f2).getChannel();

                        dstChannel.transferFrom(srcChannel, 0, srcChannel.size());
                        srcChannel.close();
                        dstChannel.close();
                    } catch (IOException e) {
                    }

                    return context.runtime.nil;
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("Takes one string argument and creates a directory with that name. It also takes an optional second argument. If it's true, will try to create all necessary directories inbetween. Default is false. Will signal a condition if the directory already exists, or if there's a file with that name.", new JavaMethod("createDirectory!") {
                private final DefaultArgumentsDefinition ARGUMENTS = DefaultArgumentsDefinition
                    .builder()
                    .withRequiredPositional("directoryName")
                    .withOptionalPositional("createPath", "false")
                    .getArguments();

                @Override
                public DefaultArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject method, IokeObject context, IokeObject message, Object on) throws ControlFlow {
                    List<Object> args = new ArrayList<Object>();
                    getArguments().getEvaluatedArguments(context, message, on, args, new HashMap<String, Object>());

                    String name = Text.getText(args.get(0));
                    File f = null;
                    if(IokeSystem.isAbsoluteFileName(name)) {
                        f = new File(name);
                    } else {
                        f = new File(context.runtime.getCurrentWorkingDirectory(), name);
                    }

                    final Runtime runtime = context.runtime;
                    if(f.exists()) {
                        String msg = null;
                        if(f.isFile()) {
                            msg = "Can't create directory '" + name + "' since there already exists a file with that name";
                        } else {
                            msg = "Can't create directory '" + name + "' since there already exists a directory with that name";
                        }

                        final IokeObject condition = IokeObject.as(IokeObject.getCellChain(runtime.condition, 
                                                                                           message, 
                                                                                           context, 
                                                                                           "Error", 
                                                                                           "IO"), context).mimic(message, context);
                        condition.setCell("message", message);
                        condition.setCell("context", context);
                        condition.setCell("receiver", on);
                        condition.setCell("text", runtime.newText(msg));

                        runtime.withReturningRestart("ignore", context, new RunnableWithControlFlow() {
                                public void run() throws ControlFlow {
                                    runtime.errorCondition(condition);
                                }});
                    }

                    if(args.size() > 1 && IokeObject.isTrue(args.get(1))) {
                        f.mkdirs();
                    } else {
                        f.mkdir();
                    }

                    return context.runtime.nil;
                }
            }));

        obj.registerMethod(runtime.newJavaMethod("Takes one string argument and removes a directory with that name. Will signal a condition if the directory doesn't exist, or if there's a file with that name.", new JavaMethod("removeDirectory!") {
                private final DefaultArgumentsDefinition ARGUMENTS = DefaultArgumentsDefinition
                    .builder()
                    .withRequiredPositional("directoryName")
                    .getArguments();

                @Override
                public DefaultArgumentsDefinition getArguments() {
                    return ARGUMENTS;
                }

                @Override
                public Object activate(IokeObject method, IokeObject context, IokeObject message, Object on) throws ControlFlow {
                    List<Object> args = new ArrayList<Object>();
                    getArguments().getEvaluatedArguments(context, message, on, args, new HashMap<String, Object>());

                    String name = Text.getText(args.get(0));
                    File f = null;
                    if(IokeSystem.isAbsoluteFileName(name)) {
                        f = new File(name);
                    } else {
                        f = new File(context.runtime.getCurrentWorkingDirectory(), name);
                    }

                    final Runtime runtime = context.runtime;
                    if(!f.exists() || f.isFile()) {
                        String msg = null;
                        if(f.isFile()) {
                            msg = "Can't remove directory '" + name + "' since it is a file";
                        } else {
                            msg = "Can't remove directory '" + name + "' since it doesn't exist";
                        }

                        final IokeObject condition = IokeObject.as(IokeObject.getCellChain(runtime.condition, 
                                                                                           message, 
                                                                                           context, 
                                                                                           "Error", 
                                                                                           "IO"), context).mimic(message, context);
                        condition.setCell("message", message);
                        condition.setCell("context", context);
                        condition.setCell("receiver", on);
                        condition.setCell("text", runtime.newText(msg));

                        runtime.withReturningRestart("ignore", context, new RunnableWithControlFlow() {
                                public void run() throws ControlFlow {
                                    runtime.errorCondition(condition);
                                }});
                    }

                    f.delete();

                    return context.runtime.nil;
                }
            }));
    }
}// FileSystem

File: src/ikc/main/Ioke.Lang/FileSystem.cs

namespace Ioke.Lang {
    using System.Collections;
    using System.Collections.Generic;
    using System.IO;

    using Ioke.Lang.Util;

    public class FileSystem {
        public static IList<string> Glob(Runtime runtime, string text) {
            return runtime.globber.PushGlob(runtime.CurrentWorkingDirectory, text, 0);
        }

        public class IokeFile : IokeIO {
            //            private FileInfo file;

            public IokeFile(FileInfo file) : base(null, null) {
                //                this.file = file;

                try {
                    if(file != null) {
                        this.writer = new StreamWriter(file.OpenWrite());
                    }
                } catch(IOException) {
                }
            }

            public override void Init(IokeObject obj) {
                Runtime runtime = obj.runtime;

                obj.Kind = "FileSystem File";

                obj.RegisterMethod(runtime.NewNativeMethod("Closes any open stream to this file",
                                                           new TypeCheckingNativeMethod.WithNoArguments("close", obj,
                                                                                                        (method, on, args, keywords, context, message) => {
                                                                                                            try {
                                                                                                                TextWriter writer = IokeIO.GetWriter(on);
                                                                                                                if(writer != null) {
                                                                                                                    writer.Close();
                                                                                                                }
                                                                                                            } catch(IOException) {
                                                                                                            }
                                                                                                            return context.runtime.nil;
                                                                                                        })));
            }
        }

        public static void Init(IokeObject obj) {
            Runtime runtime = obj.runtime;
            obj.Kind = "FileSystem";

            IokeObject file = new IokeObject(runtime, "represents a file in the file system", new IokeFile(null));
            file.MimicsWithoutCheck(runtime.Io);
            file.Init();
            obj.RegisterCell("File", file);

            obj.RegisterMethod(runtime.NewNativeMethod("Tries to interpret the given arguments as strings describing file globs, and returns an array containing the result of applying these globs.",
                                                       new NativeMethod("[]", DefaultArgumentsDefinition.builder()
                                                                        .WithRest("globTexts")
                                                                        .Arguments,
                                                                        (method, context, message, on, outer) => {
                                                                            var args = new SaneArrayList();
                                                                            outer.ArgumentsDefinition.GetEvaluatedArguments(context, message, on, args, new SaneDictionary<string, object>());

                                                                            var dirs = FileSystem.Glob(context.runtime, IokeSystem.WithReplacedHomeDirectory(Text.GetText(args[0])));
                                                                            var result = new SaneArrayList();
                                                                            foreach(string s in dirs) {
                                                                                result.Add(context.runtime.NewText(s));
                                                                            }
                                                                            return context.runtime.NewList(result);
                                                                        })));

            obj.RegisterMethod(runtime.NewNativeMethod("Takes one string argument and returns true if it's the relative or absolute name of a directory, and false otherwise.",
                                                       new NativeMethod("directory?", DefaultArgumentsDefinition.builder()
                                                                        .WithRequiredPositional("directoryName")
                                                                        .Arguments,
                                                                        (method, context, message, on, outer) => {
                                                                            var args = new SaneArrayList();
                                                                            outer.ArgumentsDefinition.GetEvaluatedArguments(context, message, on, args, new SaneDictionary<string, object>());

                                                                            string name = IokeSystem.WithReplacedHomeDirectory(Text.GetText(args[0]));
                                                                            DirectoryInfo f = null;
                                                                            if(IokeSystem.IsAbsoluteFileName(name)) {
                                                                                f = new DirectoryInfo(name);
                                                                            } else {
                                                                                f = new DirectoryInfo(Path.Combine(context.runtime.CurrentWorkingDirectory, name));
                                                                            }

                                                                            return f.Exists ? context.runtime.True : context.runtime.False;
                                                                        })));

            obj.RegisterMethod(runtime.NewNativeMethod("Takes one string argument that should be a file name, and returns a text of the contents of this file.",
                                                       new NativeMethod("readFully", DefaultArgumentsDefinition.builder()
                                                                        .WithRequiredPositional("fileName")
                                                                        .Arguments,
                                                                        (method, context, message, on, outer) => {
                                                                            var args = new SaneArrayList();
                                                                            outer.ArgumentsDefinition.GetEvaluatedArguments(context, message, on, args, new SaneDictionary<string, object>());

                                                                            string name = IokeSystem.WithReplacedHomeDirectory(Text.GetText(args[0]));
                                                                            FileInfo f = null;
                                                                            if(IokeSystem.IsAbsoluteFileName(name)) {
                                                                                f = new FileInfo(name);
                                                                            } else {
                                                                                f = new FileInfo(Path.Combine(context.runtime.CurrentWorkingDirectory, name));
                                                                            }

                                                                            return context.runtime.NewText(File.ReadAllText(f.FullName));
                                                                        })));

            obj.RegisterMethod(runtime.NewNativeMethod("Takes one string argument and returns true if it's the relative or absolute name of a file, and false otherwise.",
                                                       new NativeMethod("file?", DefaultArgumentsDefinition.builder()
                                                                        .WithRequiredPositional("fileName")
                                                                        .Arguments,
                                                                        (method, context, message, on, outer) => {
                                                                            var args = new SaneArrayList();
                                                                            outer.ArgumentsDefinition.GetEvaluatedArguments(context, message, on, args, new SaneDictionary<string, object>());

                                                                            string name = IokeSystem.WithReplacedHomeDirectory(Text.GetText(args[0]));
                                                                            FileInfo f = null;
                                                                            if(IokeSystem.IsAbsoluteFileName(name)) {
                                                                                f = new FileInfo(name);
                                                                            } else {
                                                                                f = new FileInfo(Path.Combine(context.runtime.CurrentWorkingDirectory, name));
                                                                            }

                                                                            return f.Exists ? context.runtime.True : context.runtime.False;
                                                                        })));

            obj.RegisterMethod(runtime.NewNativeMethod("Takes one string argument and returns true if it's the relative or absolute name of something that exists.",
                                                       new NativeMethod("exists?", DefaultArgumentsDefinition.builder()
                                                                        .WithRequiredPositional("entryName")
                                                                        .Arguments,
                                                                        (method, context, message, on, outer) => {
                                                                            var args = new SaneArrayList();
                                                                            outer.ArgumentsDefinition.GetEvaluatedArguments(context, message, on, args, new SaneDictionary<string, object>());

                                                                            string name = IokeSystem.WithReplacedHomeDirectory(Text.GetText(args[0]));
                                                                            string nx = null;
                                                                            if(IokeSystem.IsAbsoluteFileName(name)) {
                                                                                nx = name;
                                                                            } else {
                                                                                nx = Path.Combine(context.runtime.CurrentWorkingDirectory, name);
                                                                            }

                                                                            return (new FileInfo(nx).Exists || new DirectoryInfo(nx).Exists) ? context.runtime.True : context.runtime.False;
                                                                        })));

            obj.RegisterMethod(runtime.NewNativeMethod("Takes one string argument that should be the path of a file or directory, and returns the parent of it - or nil if there is no parent.",
                                                       new NativeMethod("parentOf", DefaultArgumentsDefinition.builder()
                                                                        .WithRequiredPositional("entryName")
                                                                        .Arguments,
                                                                        (method, context, message, on, outer) => {
                                                                            var args = new SaneArrayList();
                                                                            outer.ArgumentsDefinition.GetEvaluatedArguments(context, message, on, args, new SaneDictionary<string, object>());

                                                                            string name = Text.GetText(args[0]);
                                                                            string nx;
                                                                            if(IokeSystem.IsAbsoluteFileName(name)) {
                                                                                nx = name;
                                                                            } else {
                                                                                nx = Path.Combine(context.runtime.CurrentWorkingDirectory, name);
                                                                            }

                                                                            string parent = Path.GetDirectoryName(nx);
                                                                            if(parent == null) {
                                                                                return context.runtime.nil;
                                                                            }

                                                                            string cwd = context.runtime.CurrentWorkingDirectory;

                                                                            if(!IokeSystem.IsAbsoluteFileName(name) && parent.Equals(cwd)) {
                                                                                return context.runtime.nil;
                                                                            }

                                                                            if(parent.StartsWith(cwd)) {
                                                                                parent = parent.Substring(cwd.Length+1);
                                                                            }

                                                                            return context.runtime.NewText(parent);
                                                                        })));

            obj.RegisterMethod(runtime.NewNativeMethod("Takes a file name and a lexical block - opens the file, ensures that it exists and then yields the file to the block. Finally it closes the file after the block has finished executing, and then returns the result of the block.",
                                                       new NativeMethod("withOpenFile", DefaultArgumentsDefinition.builder()
                                                                        .WithRequiredPositional("fileName")
                                                                        .WithRequiredPositional("code")
                                                                        .Arguments,
                                                                        (method, context, message, on, outer) => {
                                                                            var args = new SaneArrayList();
                                                                            outer.ArgumentsDefinition.GetEvaluatedArguments(context, message, on, args, new SaneDictionary<string, object>());

                                                                            string name = IokeSystem.WithReplacedHomeDirectory(Text.GetText(args[0]));
                                                                            FileInfo f = null;
                                                                            if(IokeSystem.IsAbsoluteFileName(name)) {
                                                                                f = new FileInfo(name);
                                                                            } else {
                                                                                f = new FileInfo(Path.Combine(context.runtime.CurrentWorkingDirectory, name));
                                                                            }

                                                                            try {
                                                                                if(!f.Exists) {
                                                                                    using(FileStream fs = File.Create(f.FullName)) {
                                                                                    }
                                                                                }
                                                                            } catch(IOException) {
                                                                            }

                                                                            IokeObject ff = context.runtime.NewFile(context, f);
                                                                            object result = context.runtime.nil;

                                                                            try {
                                                                                result = ((Message)IokeObject.dataOf(context.runtime.callMessage)).SendTo(context.runtime.callMessage, context, args[1], ff);
                                                                            } finally {
                                                                                ((Message)IokeObject.dataOf(context.runtime.closeMessage)).SendTo(context.runtime.closeMessage, context, ff);
                                                                            }

                                                                            return result;
                                                                        })));

            obj.RegisterMethod(runtime.NewNativeMethod("Copies a file. Takes two text arguments, where the first is the name of the file to copy and the second is the name of the destination. If the destination is a directory, the file will be copied with the same name, and if it's a filename, the file will get a new name",
                                                       new NativeMethod("copyFile", DefaultArgumentsDefinition.builder()
                                                                        .WithRequiredPositional("fileName")
                                                                        .WithRequiredPositional("destination")
                                                                        .Arguments,
                                                                        (method, context, message, on, outer) => {
                                                                            var args = new SaneArrayList();
                                                                            outer.ArgumentsDefinition.GetEvaluatedArguments(context, message, on, args, new SaneDictionary<string, object>());

                                                                            string name = Text.GetText(args[0]);
                                                                            FileInfo f = null;
                                                                            if(IokeSystem.IsAbsoluteFileName(name)) {
                                                                                f = new FileInfo(name);
                                                                            } else {
                                                                                f = new FileInfo(Path.Combine(context.runtime.CurrentWorkingDirectory, name));
                                                                            }

                                                                            string name2 = Text.GetText(args[1]);
                                                                            string nx = null;
                                                                            if(IokeSystem.IsAbsoluteFileName(name2)) {
                                                                                nx = name2;
                                                                            } else {
                                                                                nx = Path.Combine(context.runtime.CurrentWorkingDirectory, name2);
                                                                            }

                                                                            if(new DirectoryInfo(nx).Exists) {
                                                                                nx = Path.Combine(nx, f.Name);
                                                                            }


                                                                            try {
                                                                                File.Copy(f.FullName, nx, true);
                                                                            } catch (IOException) {
                                                                            }

                                                                            return context.runtime.nil;
                                                                        })));

            obj.RegisterMethod(runtime.NewNativeMethod("Takes one string argument and creates a directory with that name. It also takes an optional second argument. If it's true, will try to create all necessary directories inbetween. Default is false. Will signal a condition if the directory already exists, or if there's a file with that name.",
                                                       new NativeMethod("createDirectory!", DefaultArgumentsDefinition.builder()
                                                                        .WithRequiredPositional("directoryName")
                                                                        .WithOptionalPositional("createPath", "false")
                                                                        .Arguments,
                                                                        (method, context, message, on, outer) => {
                                                                            var args = new SaneArrayList();
                                                                            outer.ArgumentsDefinition.GetEvaluatedArguments(context, message, on, args, new SaneDictionary<string, object>());

                                                                            string name = IokeSystem.WithReplacedHomeDirectory(Text.GetText(args[0]));
                                                                            DirectoryInfo f = null;
                                                                            if(IokeSystem.IsAbsoluteFileName(name)) {
                                                                                f = new DirectoryInfo(name);
                                                                            } else {
                                                                                f = new DirectoryInfo(Path.Combine(context.runtime.CurrentWorkingDirectory, name));
                                                                            }

                                                                            if(f.Exists || new FileInfo(f.FullName).Exists) {
                                                                                string msg = null;
                                                                                if(f.Exists) {
                                                                                    msg = "Can't create directory '" + name + "' since there already exists a directory with that name";
                                                                                } else {
                                                                                    msg = "Can't create directory '" + name + "' since there already exists a file with that name";
                                                                                }

                                                                                IokeObject condition = IokeObject.As(IokeObject.GetCellChain(runtime.Condition,
                                                                                                                                             message,
                                                                                                                                             context,
                                                                                                                                             "Error",
                                                                                                                                             "IO"), context).Mimic(message, context);
                                                                                condition.SetCell("message", message);
                                                                                condition.SetCell("context", context);
                                                                                condition.SetCell("receiver", on);
                                                                                condition.SetCell("text", runtime.NewText(msg));

                                                                                runtime.WithReturningRestart("ignore", context, ()=>{runtime.ErrorCondition(condition);});
                                                                            }

                                                                            Directory.CreateDirectory(f.FullName);
                                                                            return context.runtime.nil;
                                                                        })));

            obj.RegisterMethod(runtime.NewNativeMethod("Takes one string argument and removes a directory with that name. Will signal a condition if the directory doesn't exist, or if there's a file with that name.",
                                                       new NativeMethod("removeDirectory!", DefaultArgumentsDefinition.builder()
                                                                        .WithRequiredPositional("directoryName")
                                                                        .Arguments,
                                                                        (method, context, message, on, outer) => {
                                                                            var args = new SaneArrayList();
                                                                            outer.ArgumentsDefinition.GetEvaluatedArguments(context, message, on, args, new SaneDictionary<string, object>());

                                                                            string name = IokeSystem.WithReplacedHomeDirectory(Text.GetText(args[0]));
                                                                            string nf = null;
                                                                            if(IokeSystem.IsAbsoluteFileName(name)) {
                                                                                nf = name;
                                                                            } else {
                                                                                nf = Path.Combine(context.runtime.CurrentWorkingDirectory, name);
                                                                            }

                                                                            if(!(new DirectoryInfo(nf).Exists) || new FileInfo(nf).Exists) {
                                                                                string msg = null;
                                                                                if(!(new DirectoryInfo(nf).Exists)) {
                                                                                    msg = "Can't remove directory '" + name + "' since it doesn't exist";
                                                                                } else {
                                                                                    msg = "Can't remove directory '" + name + "' since it is a file";
                                                                                }

                                                                                IokeObject condition = IokeObject.As(IokeObject.GetCellChain(runtime.Condition,
                                                                                                                                             message,
                                                                                                                                             context,
                                                                                                                                             "Error",
                                                                                                                                             "IO"), context).Mimic(message, context);
                                                                                condition.SetCell("message", message);
                                                                                condition.SetCell("context", context);
                                                                                condition.SetCell("receiver", on);
                                                                                condition.SetCell("text", runtime.NewText(msg));

                                                                                runtime.WithReturningRestart("ignore", context, ()=>{runtime.ErrorCondition(condition);});
                                                                            }

                                                                            Directory.Delete(nf);
                                                                            return context.runtime.nil;
                                                                        })));

            obj.RegisterMethod(runtime.NewNativeMethod("Takes one string argument and removes a file with that name. Will signal a condition if the file doesn't exist, or if there's a directory with that name.",
                                                       new NativeMethod("removeFile!", DefaultArgumentsDefinition.builder()
                                                                        .WithRequiredPositional("fileName")
                                                                        .Arguments,
                                                                        (method, context, message, on, outer) => {
                                                                            var args = new SaneArrayList();
                                                                            outer.ArgumentsDefinition.GetEvaluatedArguments(context, message, on, args, new SaneDictionary<string, object>());

                                                                            string name = IokeSystem.WithReplacedHomeDirectory(Text.GetText(args[0]));
                                                                            string nf = null;
                                                                            if(IokeSystem.IsAbsoluteFileName(name)) {
                                                                                nf = name;
                                                                            } else {
                                                                                nf = Path.Combine(context.runtime.CurrentWorkingDirectory, name);
                                                                            }

                                                                            if(!(new FileInfo(nf).Exists) || new DirectoryInfo(nf).Exists) {
                                                                                string msg = null;
                                                                                if(!(new FileInfo(nf).Exists)) {
                                                                                    msg = "Can't remove file '" + name + "' since it doesn't exist";
                                                                                } else {
                                                                                    msg = "Can't remove file '" + name + "' since it is a directory";
                                                                                }

                                                                                IokeObject condition = IokeObject.As(IokeObject.GetCellChain(runtime.Condition,
                                                                                                                                             message,
                                                                                                                                             context,
                                                                                                                                             "Error",
                                                                                                                                             "IO"), context).Mimic(message, context);
                                                                                condition.SetCell("message", message);
                                                                                condition.SetCell("context", context);
                                                                                condition.SetCell("receiver", on);
                                                                                condition.SetCell("text", runtime.NewText(msg));

                                                                                runtime.WithReturningRestart("ignore", context, ()=>{runtime.ErrorCondition(condition);});
                                                                            }

                                                                            File.Delete(nf);

                                                                            return context.runtime.nil;
                                                                        })));
        }
    }
}
